cpp/tests/link_prediction/similarity_compare.cpp

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governin_from_mtxg permissions and
+ * limitations under the License.
+ */
+
+#include <link_prediction/similarity_compare.hpp>
+
+#include <thrust/execution_policy.h>
+#include <thrust/sort.h>
+
+// FIXME:  To support weighted variants we need to compute the minimum weight of edge edge going to
+// the intersection
+//         and the maximum weight of each edge otherwise.  Need to think about how that works.
+//         Probably need to implement a custom set intersection loop rather than using the stl
+//         version, since then we can operate on each element appropriately.  That would mean
+//         adapting this class.
+template <typename T>
+struct intersection_count_t {
+  size_t count{0};
+  using value_type = T;
+
+  void push_back(T const&) { ++count; }
+};
+
+namespace cugraph {
+namespace test {
+
+template <typename vertex_t, typename weight_t, typename test_t>
+void similarity_compare(vertex_t num_vertices,
+                        std::vector<vertex_t>&& src,
+                        std::vector<vertex_t>&& dst,
+                        std::optional<std::vector<weight_t>>&& wgt,
+                        std::vector<vertex_t>&& result_src,
+                        std::vector<vertex_t>&& result_dst,
+                        std::vector<weight_t>&& result_score,
+                        test_t const& test_functor)
+{
+  ASSERT_TRUE(num_vertices > 1) << "number of vertices expected to be non-zero";
+  ASSERT_EQ(src.size(), dst.size());
+  ASSERT_EQ(result_src.size(), result_dst.size());
+  ASSERT_EQ(result_src.size(), result_score.size());
+
+  thrust::sort(thrust::host,
+               thrust::make_zip_iterator(src.begin(), dst.begin()),
+               thrust::make_zip_iterator(src.end(), dst.end()));
+  thrust::sort(
+    thrust::host,
+    thrust::make_zip_iterator(result_src.begin(), result_dst.begin(), result_score.begin()),
+    thrust::make_zip_iterator(result_src.end(), result_dst.end(), result_score.end()));
+
+  size_t result_pos = 0;
+
+  // Iterate over all (u,v) pairs
```

Comment:
Should we do this??? This can be very expensive (even though this is just for testing).

Reply:
(`src`, `dst`, `wgt`) is a graph edge list, right? I guess `result_src` and `result_dst` here are misnomers. They are first and second elements of vertex pairs, right?

If we take the CSR graph, and we can iterate over vertex pairs (`result_src` and `result_dst`), and for the first and second elements of the pair, we can extract neighbor lists from the CSR graph. Then, we can compute the expected score from the host, and compute this number with the `result_score`. I guess this will be much faster (and also easier to understand). Am I missing something?

Reply:
Several thoughts mixed in this reply.
* Yes, the tuple (`src`, `dst`, `wgt`) is the graph edge list.  I could make that a tuple called graph edge list, which would make things more semantically clear.  Perhaps that will be easier to understand.
* With the above change, `result_src` and `result_dst` will become `result_vertex_pairs`

If we make the change to compute the 2-hop neighbors, I can change this to the following:
* Add a separate test for the 2-hop neighbor function and validate it works properly
* In this test, pass in the 2-hop neighbors, and just iterate over those and directly compute the scores

Ultimately I was going to wait until I started testing to consider other optimizations to this function (other than the above).  If we're testing a full set of 2-hop neighbors, I'm not sure that extracting everything on demand will actually be faster than just operating on the fully-extracted graph.

Reply:
Perhaps we can add a parameter that allows us to test a subset of the edges rather than the whole.  For smaller graphs we can test everything, if you want to test on a bigger graph we just validate some of the vertex pairs.

Reply:
Now I see why you did this. You want to test this for every possible vertex pair. Yeah... this might be great, but can be too expensive for non-tiny graphs. I guess testing for two-hop neighbors with limited K (for high-degree vertices, pick only up to K neighbors, similar to neighbor sampling) might be sufficient.